### PR TITLE
desktop/notification: add unit tests

### PR DESCRIPTION
**Pull request by zyga**
This is based on https://github.com/snapcore/snapd/pull/9410

Everything but is now fully unit tested.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: desktop/notification/fdo_test.go [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	_, err := srv.ServerCapabilities()
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+func (s *fdoSuite) TestSendNotificationSuccess(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
```
Comment (stolowski) [inline desktop/notification/fdo_test.go:201]: I think an `if` would be more clear, unless you plan to add more cases?

File: desktop/notification/fdo_test.go [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	_, err := srv.SendNotification(&notification.Message{})
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+func (s *fdoSuite) TestCloseNotificationSuccess(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
```
Comment (stolowski) [inline desktop/notification/fdo_test.go:266]: Same, maybe just `if n != 0 { ... error case ... }`

File: desktop/notification/fdo_test.go [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	err := srv.CloseNotification(notification.ID(42))
+	c.Assert(err, IsNil)
+}
+
+func (s *fdoSuite) TestCloseNotificationError(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
```
Comment (stolowski) [inline desktop/notification/fdo_test.go:288]: And here.

**Comment by zyga** [comment]
@stolowski I used the `switch n` pattern that we have in couple of other places but I can reduce the simpler cases.

**Review by stolowski — APPROVED** [review]
Thank you. LGTM. As for the switch vs if, it's fine; I think I would slightly prefer error cases to be moved inside the switch statements as `default:` clauses, but no biggie.

File: desktop/notification/fdo_test.go [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+			return []*dbus.Message{response}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	err := srv.CloseNotification(notification.ID(42))
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+type testObserver struct {
+	notificationClosed func(notification.ID, notification.CloseReason) error
```
Comment (mvo5) [inline desktop/notification/fdo_test.go:300]: (nitpick) I think I would have done something like:
```
type testObserver struct {
	notificationClosed []struct {
		notification.ID
		notification.CloseReason
	}
	notificationClosedErr error

	actionInvoked []struct {
		notification.ID
		string
	}
	actionInvokedErr error
}

func (o *testObserver) NotificationClosed(id notification.ID, reason notification.CloseReason) error {
	o.notificationClosed = append(o.notificationClosed, struct {
		notification.ID
		notification.CloseReason
	}{id, reason})

	return o.notificationClosedErr
}
...
```
and then just "c.Check(DeepEquals)" on the testObserver.notificationClosed. Slightly more compact in the individual tests but probably just a matter of preferences.

File: desktop/notification/fdo_test.go [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+		case 1:
+			s.checkRemoveMatchRequest(c, msg)
+			response := s.removeMatchResponse(c, msg)
+			return []*dbus.Message{response}, nil
+		default:
+			return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+		}
+	})
+	err := srv.ObserveNotifications(context.TODO(), &testObserver{
+		actionInvoked: func(id notification.ID, actionKey string) error {
```
Comment (mvo5) [inline desktop/notification/fdo_test.go:492]: (nitpick) this (and similar ones) might increase a counter to ensure the action was only invoked once.

File: desktop/notification/fdo_test.go [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+
+	err = notification.ProcessSignal(&dbus.Signal{
+		Name: "org.freedesktop.Notifications.ActionInvoked",
+		Body: []interface{}{true, "action-key"},
+	}, &testObserver{})
+	c.Assert(err, ErrorMatches, "cannot process ActionInvoked signal: expected first body element to be uint32, got bool")
+}
+
+func (s *fdoSuite) TestProcessNotificationClosedSignalSuccess(c *C) {
+	called := false
```
Comment (mvo5) [inline desktop/notification/fdo_test.go:559]: (nitpick) I would use a counter here too.

**Review by mvo5 — APPROVED** [review]
Looks great, nitpicks inside but feel free to ignore. This is probably the best tested fdo notification implementation.

**Comment by zyga** [comment]
Thank you for the review! I will do a follow-up as there is more that I want to push here.

**Pull request merged**